lib/benchmark/benchmark.go

```diff
@@ -0,0 +1,127 @@
+package benchmark
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/HdrHistogram/hdrhistogram-go"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/trace"
+	logrus "github.com/sirupsen/logrus"
+)
+
+const (
+	// PauseTimeBetweenBenchmarks is the time to pause between each benchmark
+	PauseTimeBetweenBenchmarks = time.Second * 5
+)
+
+// Generator provides a standardized way to get successive benchmarks from a consistent interface
+type Generator interface {
+	// GetBenchmark returns the benchmark config for the current generation.
+	// Returns context, Config and error
+	GetBenchmark() (context.Context, Config, error)
+	// Benchmark runs the benchmark of receiver type
+	// return an array of Results that contain information about the generations
+	Benchmark(context.Context, []string, *client.TeleportClient) ([]*Result, error)
+}
+
+// Run is used to run the benchmarks, it is given a generator, command to run,
+// a host, host login, and proxy. If host login or proxy is an empty string, it will
+// use the default login
+func Run(ctx context.Context, lg Linear, cmd, host, login, proxy string) ([]Result, error) {
+	var result Result
+	var results []Result
+	logrus.SetLevel(logrus.ErrorLevel)
+	c := strings.Split(cmd, " ")
+	lg.config.Command = c
+	lg.Threads = 1
+	sleep := false
+	if err := validateConfig(lg); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	tc, err := makeTeleportClient(host, login, proxy)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	for {
+		if sleep {
+			time.Sleep(PauseTimeBetweenBenchmarks)
+		}
```

Comment:
the `sleep` flag exists to avoid sleeping before the first run, right?
if so, move `time.Sleep` to the end of the loop and remove the `sleep` var entirely

Reply:
If I were to do this, the last run would sleep for 5 seconds before being finished.